### PR TITLE
materialized: switch to cockroach in docker image

### PR DESCRIPTION
**Pull request by mjibson**
This is what cloud deployments use, so we should test with it.

Additionally, we hope it will fix the recent test flakes that have become common.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: misc/images/materialized/Dockerfile [hunk 1]
```diff
@@ -9,25 +9,22 @@
...
     && mkdir /mzdata \
-    && chown materialize /mzdata
+    && mkdir /cockroach-data \
+    && chown materialize /mzdata /cockroach-data
 
-COPY pg_hba.conf /etc/postgresql/14/materialize/pg_hba.conf
+ARG COCKROACH_VERSION=22.1.5
+
+RUN curl https://binaries.cockroachdb.com/cockroach-v${COCKROACH_VERSION}.linux-amd64.tgz | tar -xz \
+    && cp -i cockroach-v${COCKROACH_VERSION}.linux-amd64/cockroach /usr/local/bin/
```
Comment (benesch) [inline misc/images/materialized/Dockerfile:27]: `-i` (interactive) seems weird here?

**Review by benesch — APPROVED** [review]
🙇🏽

File: misc/images/materialized/Dockerfile (hunk 1 above)
Comment (mjibson) [reply misc/images/materialized/Dockerfile:27]: :shrug: copied directly from https://www.cockroachlabs.com/docs/stable/install-cockroachdb-linux.html

Comment (mjibson) [reply misc/images/materialized/Dockerfile]: This is an argument I didn't know what it did. Seems like should be removed in docker, yeah. Done.

**Comment by mjibson** [comment]
RIP

> A Tokio 1.x context was found, but it is being shutdown.

**Comment by benesch** [comment]
> RIP
> 
> > A Tokio 1.x context was found, but it is being shutdown.
> 
> (This error occurred and I had to click retry.)

That is expected! My hope is that this fixes all the _other_ hangs we've been seeing in CI. I have a PR out soon for the reactor shutdown issue.

**Comment by mjibson** [comment]
Just also saw:

> deadline has elapsed

**Comment by mjibson** [comment]
I've seen deadline elapsed about 10 times now, but no stash error? Maybe that's an improvement?

**Review by benesch — COMMENTED** [review]
Very spooky error from the Postgres CDC tests:

```
pg-cdc-materialized-1  | thread 'timely:work-0' panicked at 'Postgres connection unexpectedly missing secrets: No such file or directory (os error 2)
```

**Comment by mjibson** [comment]
Now seeing:

```
error: preparing query failed: db error: ERROR: unknown catalog item 'char_type_quoted_view': ERROR: unknown catalog item 'char_type_quoted_view'
```

```
check-from-current_source-database-schema.td:10:1: error: non-matching rows: expected:
[["materialize"], ["some database"]]
got:
[["materialize"]]
Poor diff:
- "some database"
```

The changed Dockerfile in this PR had `-pgoption max_connections=5000`, and cockroach doesn't have low connection limits. That we are seeing the same problems suggests this wasn't a connection issue. It's feeling more like a "tests running concurrently and messing each other up" issue now.

**Comment by benesch** [comment]
I agree that it's not low connection limits! Something seems really fishy here. I'm not sure it's concurrency, though: many of these failing tests run on their own build hosts and only do one thing at a time?

**Comment by mjibson** [comment]
Just got a `error: deadline has elapsed` locally when running `bin/mzcompose --find pg-cdc run default` so not a buildkite issue?

**Comment by benesch** [comment]
Yeah, I suspect a real bug! Either in persist or timestamp selection, I
would guess. Would be illuminating to connect to the materialized in the
test that failed and run EXPLAIN TIMESTAMP on the query that hung.

**Comment by philip-stoev** [comment]
@benesch are we all in on Cockroach? Because we have a separate Postgres mzcompose service that also needs to be fixed , along with the tests that use it. The test/cloudtest thing also uses postgres ...

**Comment by benesch** [comment]
Oh, I think I see the bug. This needs to create a volume for /cockroach-data so that the stash survives a restart.

**Comment by mjibson** [comment]
This is currently blocked because a recent migration test uses postgres in a hard coded version, but if we switch to cockroach then the upgrade test fails.

**Pull request merged**